src/Microsoft.DotNet.Darc/src/DarcLib/VirtualMonoRepo/VmrScanner.cs

```diff
@@ -79,12 +82,59 @@ private async Task<string[]> ScanRepository(SourceMapping sourceMapping, Cancell
 
         foreach (var file in files)
         {
-            _logger.LogWarning("File {file} is cloaked but present in the VMR", file);
+            _logger.LogWarning("File in {file} is cloaked but present in the VMR", file);
         }
 
         return files;
     }
 
-    private static string ExcludeFile(string file) => $":(attr:!{VmrInfo.KeepAttribute}){file}";
+    private static string ExcludeFile(string file, string gitAttribute) => $":(attr:!{gitAttribute}){file}";
+
+    public async Task<List<string>> ListBinaryFiles(CancellationToken cancellationToken)
+    {
+        await _dependencyTracker
+            .InitializeSourceMappings(_vmrInfo.VmrPath / VmrInfo.SourcesDir / VmrInfo.SourceMappingsFileName);
+
+        var binaryFiles = new List<string>();
+        var taskList = new List<Task<IEnumerable<string>>>();
+
+        foreach (var mapping in _dependencyTracker.Mappings)
+        {
+            taskList.Add(ScanRepositoryForBinaryFiles(mapping, cancellationToken));
+        }
+
+        await Task.WhenAll(taskList);
+
+        return taskList.SelectMany(task => task.Result).ToList();
+    }
+
+    private async Task<IEnumerable<string>> ScanRepositoryForBinaryFiles(SourceMapping sourceMapping, CancellationToken cancellationToken)
+    {
+        cancellationToken.ThrowIfCancellationRequested();
+        _logger.LogInformation("Scanning {repository} repository", sourceMapping.Name);
+        var args = new string[]{
+            "diff",
+            Constants.EmptyGitObject,
+            "--numstat",
+            ExcludeFile(_vmrInfo.GetRepoSourcesPath(sourceMapping), VmrInfo.VmrIgnoreBinaryAttribute)
+        };
+
+        var ret = await _processManager.ExecuteGit(_vmrInfo.VmrPath, args.ToArray(), cancellationToken);
+
+        ret.ThrowIfFailed($"Failed to scan the {sourceMapping.Name} repository");
+
+        var files = ret.StandardOutput
+            .Split(Environment.NewLine, StringSplitOptions.RemoveEmptyEntries)
+            .Where(line => line.StartsWith(_binaryFileMarker))
+            .Select(line => line.Split("\t").Last());
+
+
+        foreach (var file in files)
+        {
+            _logger.LogWarning("Found binary file {file} in {repo} repo", file, sourceMapping.Name);
```

Comment:
Since we're doing this in parallel, doesn't the output get mixed from different threads?

Reply:
The outputs don't have to be sorted by repo, that's true, if we're doing this this way. If that is something we really care about, we could not log them here, but have them sorted once they're all done and log all of them then.
Also, the "scanning arcade repo.." is not really useful in this scenario

Reply:
It would be helpful to have them sorted, but I also meant - can it happen 2 lines are outputted together on a single line garbled together because of race condition?

Reply:
I'd just return a sorted list and let the command itself log it